modules/gapi/test/serialization/serialization.cpp

```diff
@@ -0,0 +1,1292 @@
+    std::vector<ade::NodeHandle> nodes;
+
+    for (const auto& nh : g.nodes())
+    {
+        if (cgm.metadata(nh).get<NodeType>().t == NodeType::DATA)
+        {
+            auto d = cgm.metadata(nh).get<gimpl::Data>();
+            dataNodes[d.shape][d.rc] = nh;
+        }
+        nodes.push_back(nh);
+    }
+
+    for (const auto& nh : g.nodes())
+    {
+        if (cgm.metadata(nh).get<NodeType>().t == NodeType::OP)
+        {
+            const auto& op = cgm.metadata(nh).get<gimpl::Op>();
+
+            for (const auto& in : ade::util::indexed(op.args))
+            {
+                const auto& arg = ade::util::value(in);
+                if (arg.kind == detail::ArgKind::GOBJREF)
+                {
+                    const auto idx = ade::util::index(in);
+                    const auto rc = arg.get<gimpl::RcDesc>();
+
+                    const auto& inDataNode = dataNodes[rc.shape][rc.id];
+                    const auto& e = g.link(inDataNode, nh);
+                    gm.metadata(e).set(Input{idx});
+                }
+            }
+
+            for (const auto& out : ade::util::indexed(op.outs))
+            {
+                const auto idx = ade::util::index(out);
+                const auto rc = ade::util::value(out);
+
+                const auto& outDataNode = dataNodes[rc.shape][rc.id];
+                const auto& e = g.link(nh, outDataNode);
+                gm.metadata(e).set(Output{idx});
+            }
+        }
+    }
+    return nodes;
+}
+//} // anonymous namespace
+
+GSerialized serialize(const gimpl::GModel::ConstGraph& cg, const std::vector<ade::NodeHandle>& nodes)
+{
+    GSerialized s;
+    for (auto &nh : nodes)
+    {
+        switch (cg.metadata(nh).get<NodeType>().t)
+        {
+        case NodeType::OP:   putOp  (s, cg, nh); break;
+        case NodeType::DATA: putData(s, cg, nh); break;
+        default: util::throw_error(std::logic_error("Unknown NodeType"));
+        }
+    }
+    return s;
+}
+
+void deserialize(const serialization::GSerialized& s)
+{
+    printGSerialized(s);
+
+    // FIXME: reuse code from GModelBuilder/GModel!
+    // ObjectCounter?? (But seems we need existing mapping by shape+id)
+
+    ade::Graph g;
+
+    for (const auto& data : s.m_datas)
+    {
+        mkDataNode(g, data);
+    }
+
+    for (const auto& op : s.m_ops)
+    {
+        mkOpNode(g, op);
+    }
+
+    linkNodes(g);
+
+//  FIXME:
+//  Handle IslandModel!
+//    std::shared_ptr<ade::Graph> ig;
+//    GModel::Graph gm(g);
+//    gm.metadata().set(gimpl::IslandModel{std::move(ig)});
+
+    auto pass_ctx = ade::passes::PassContext{g};
+    ade::passes::TopologicalSort{}(pass_ctx);
+    gimpl::passes::dumpDotToFile(pass_ctx, "graph.dot");
+}
+
+/////////////////////////////////////////////////////////////////////////////////////
+//Graph dump operators
+template<typename _Tp> static inline
+SerializationStream&  operator << (SerializationStream& os, /*const*/ _Tp value)
+{
+    os << value;
+    return os;
+}
```

Comment:
This thing is here to enter the recursion, right :-)

Reply:
???

Reply:
Just have a closer look on this code.